### PR TITLE
Fixing - https://wso2.org/jira/browse/IDNCLOUD-151

### PR DESCRIPTION
**Pull request by sajithab**
Fixing mutual SSL does not work with hostname verification enabled issue

### CODE REVIEW

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import javax.net.ssl.SSLContext;
+import java.io.*;
+import java.security.*;
+import java.security.cert.CertificateException;
+import java.util.Map;
+
+public class AppMHttpClient {
+
+    private static final Log log = LogFactory.getLog(AppMHttpClient.class);
+    CloseableHttpClient httpClient = null;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: Just reference with HttpClient

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
+/*
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: Let's rename the class as MutualSSLHttpClient

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    private static String keyStoreType = "JKS";
+    private static String SecurityKeyStoreLocation = "Security.KeyStore.Location";
+    private static String SecurityKeyStorePassword = "Security.KeyStore.Password";
+    private static String SecurityKeyStoreKeyPassword = "Security.KeyStore.KeyPassword";
+    private static String SecurityTrustStoreLocation = "Security.TrustStore.Location";
+    private static String SecurityTrustStorePassword = "Security.TrustStore.Password";
+
+    public AppMHttpClient() {
+        try {
+            // Load the keystore containing the client certificate - keystore type is probably jks or pkcs12.
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: Remove un relevant comments

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            final KeyStore keyStore = KeyStore.getInstance(keyStoreType);
+            InputStream keystoreInput = new FileInputStream(new File(ServerConfiguration.getInstance()
+                    .getFirstProperty(SecurityKeyStoreLocation)));
+            keyStore.load(keystoreInput, ServerConfiguration.getInstance()
+                    .getFirstProperty(SecurityKeyStorePassword).toCharArray());
+
+            // Load the trust store, leave it null to rely on cacerts distributed with the JVM - truststore type is
+            // probably jks or pkcs12.
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: Remove un relevant comments

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        } catch (CertificateException e) {
+            log.error(e);
+        } catch (NoSuchAlgorithmException e) {
+            log.error(e);
+        } catch (UnrecoverableKeyException e) {
+            log.error(e);
+        } catch (KeyManagementException e) {
+            log.error(e);
+        }
+    }
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: Add appropriate error messages and collapse those falls under same category

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        String responseString = null;
+        try {
+            StringEntity inputMappings = new StringEntity(jsonContent);
+            inputMappings.setContentType(ApplicationJson);
+            HttpPost postMethod = new HttpPost(endPoint);
+            postMethod.setEntity(inputMappings);
+            responseString = doHttpMethod(postMethod, headers);
+        } catch (UnsupportedEncodingException e) {
+            log.error(e);
+            return null;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: redundant return

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        String responseString = null;
+        try {
+            StringEntity inputMappings  = new StringEntity(jsonContent);
+            inputMappings.setContentType(ApplicationJson);
+            HttpPut putMethod = new HttpPut(endPoint);
+            putMethod.setEntity(inputMappings);
+            responseString = doHttpMethod(putMethod, headers);
+        } catch (UnsupportedEncodingException e) {
+            log.error(e);
+            return null;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: redundant return

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            for (Map.Entry<String, String> entry : headers.getHeaderMap().entrySet()) {
+                httpMethod.addHeader(entry.getKey(), entry.getValue());
+            }
+
+            CloseableHttpResponse closeableHttpResponse = httpClient.execute(httpMethod);
+            HttpEntity entity = closeableHttpResponse.getEntity();
+            responseString = EntityUtils.toString(entity, "UTF-8");
+        } catch (IOException e) {
+            log.error(e);
+            return null;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: redundant return

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/internal/JaggeryWebComponent.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    protected void activate(ComponentContext ctxt) {
+        if (log.isDebugEnabled()){
+            log.debug("JaggeryWebComponent activated");
+        }
+    }
+
+    protected void deactivate(ComponentContext ctxt) {
+        if (log.isDebugEnabled()){
+            log.debug("HostObjectComponent deactivated");
```
Comment (malithie) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/internal/JaggeryWebComponent.java]: Add the respective component name to the message

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 1 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 2 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 3 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 4 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 5 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 6 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 7 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java (hunk 8 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/clients/AppMHttpClient.java]: done

File: components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/internal/JaggeryWebComponent.java (hunk 9 above)
Comment (sajithab) [inline components/org.wso2.carbon.identity.cloud.web.jaggery/src/main/java/org/wso2/carbon/identity/cloud/web/jaggery/internal/JaggeryWebComponent.java]: done

**Pull request merged**